core/templates/dev/head/pages/exploration_editor/editor_tab/ResponsesService.js

```diff
@@ -71,31 +62,9 @@ oppia.factory('ResponsesService', [
           stateSolutionService.savedMemento.correctAnswer !== null);
 
         if (interactionCanHaveSolution && solutionExists) {
-          var currentStateName = EditorStateService.getActiveStateName();
-          var solutionWasPreviouslyValid = (
-            ExplorationStatesService.isSolutionValid(
-              EditorStateService.getActiveStateName()));
-          var solutionIsCurrentlyValid = (
-            SolutionVerificationService.verifySolution(
-              ContextService.getExplorationId(),
-              ExplorationStatesService.getState(currentStateName),
-              stateSolutionService.savedMemento.correctAnswer));
-
-          ExplorationStatesService.updateSolutionValidity(
-            currentStateName, solutionIsCurrentlyValid);
-          ExplorationWarningsService.updateWarnings();
-
-          if (solutionIsCurrentlyValid && !solutionWasPreviouslyValid) {
-            AlertsService.addInfoMessage(INFO_MESSAGE_SOLUTION_IS_VALID);
-          } else if (!solutionIsCurrentlyValid && solutionWasPreviouslyValid) {
-            AlertsService.addInfoMessage(
-              INFO_MESSAGE_SOLUTION_IS_INVALID_FOR_CURRENT_RULE);
-          } else if (!solutionIsCurrentlyValid && !solutionWasPreviouslyValid) {
-            AlertsService.addInfoMessage(INFO_MESSAGE_SOLUTION_IS_INVALID);
-          }
+          $rootScope.$broadcast('validateSolution', true);
```

Comment:
Mm, this doesn't look right. Avoid using $rootScope.broadcast, that's too broad.

If you need the cache, maybe this is a good use for the service I suggested in StateEditorDirective.js for the readonly stuff. Or possibly you can call SolutionVerificationService directly, if it doesn't have exploration-editor-specific dependencies.

Reply:
Yeah, but the issue was that SolutionVerificationService required the state also to be passed in.

How about I modify the existing EditorStateService to also record the state that is currently active.
So, basically make it a neutral service that hold everything related to only the state without any context of the exploration? Then, maybe this same service can be reused in question editor as well.

Reply:
I think you need to decide which of these are exploration-specific things, and which of them are meant to be used in both explorations and questions. That will tell you what goes where.

For warnings, if that's an exploration-specific thing, have an onAnswerGroupsSaved() callback and call it. Let the exploration editor (outside the state editor directive) handle the work.

For solution verification, there are two things: on-the-fly solution verification, and caching solution validity by state. Seems to me like the first of these is common to both explorations and questions, and the second is an exploration-specific thing. So, if you need on-the-fly solution verification, call it directly from inside the StateEditor or any of its child directives, and make sure you do that statelessly without side-effects and without passing any additional information like exp ID or state name. And if you need solution validity caching, then after you've done the verification, call onSolutionValidityUpdated() to modify and update the cache that's used in the exploration editor page.

Does that help? Basically, the recipe is: figure out whether this is a StateEditor directive thing or whether it's specific to one of the pages. If it's the former, create helper services for the StateEditor directive that should be imported in multiple pages and that don't depend on anything that is specific to one of the pages. If it's the latter, use a callback to pass the relevant info out of the directive to the parent and let the surrounding page deal with it.

Note that some refactoring of existing services and separation of concerns might be needed (e.g. you might need to split the caching bit out of SolutionVerificationService), and that's totally fine. Pretend you were designing it from scratch and don't feel too bound by how things were separated out into different services etc. in the past, since that happened under different assumptions.

Reply:
Actually, SolutionValidity caching is fine (requires only state name which is available), it was SolutionVerification that was the issue, since it was taking the complete state as an argument, but on looking at its actual usage, it seems it only needs interaction, so I'll just set interaction in EditorStateService in the beginning (similar to the others) and use it for SolutionVerificationService.
Ok, yeah, I'll try using callbacks and remove broadcasting.
Thanks!

Reply:
Done, removed all broadcasts.